### PR TITLE
Add support for 3.5

### PR DESCRIPTION
**Pull request by duncaan**

### CODE REVIEW

File: 3.5.3/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,72 @@
+FROM openjdk:8-jre-alpine
+MAINTAINER Elisey Zanko <elisey.zanko@gmail.com>
```
Comment (31z4) [inline 3.5.3/Dockerfile]: Let's remove it. It's been depricated since Docker `v1.13.0`.

File: 3.5.3/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
```
Comment (31z4) [inline 3.5.3/Dockerfile]: Please remove empty line.

File: 3.5.3/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+ARG DISTRO_NAME=zookeeper-3.5.3-beta
+
+# Download Apache Zookeeper, verify its PGP signature, untar and clean up
+RUN set -ex; \
+    apk add --no-cache --virtual .build-deps \
+        gnupg; \
+    wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz"; \
+    wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz.asc"; \
+    wget -q "http://www.apache.org/dist/zookeeper/KEYS"; \
```
Comment (31z4) [inline 3.5.3/Dockerfile]: 👍

File: 3.5.3/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    export GNUPGHOME="$(mktemp -d)"; \
+    gpg --import KEYS ; \
+    gpg --batch --verify "$DISTRO_NAME.tar.gz.asc" "$DISTRO_NAME.tar.gz"; \
+    # TODO: the distribution given by apache is not gzipped correctly - unzip correctly when fixed
```
Comment (31z4) [inline 3.5.3/Dockerfile]: What's exactly wrong with it?

File: 3.5.3/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      "$DISTRO_NAME/docs" \
+      "$DISTRO_NAME/ivy.xml" \
+      "$DISTRO_NAME/ivysettings.xml" \
+      "$DISTRO_NAME/recipes" \
+      "$DISTRO_NAME/src" \
+      "$DISTRO_NAME/$DISTRO_NAME.jar.asc" \
+      "$DISTRO_NAME/$DISTRO_NAME.jar.md5" \
+      "$DISTRO_NAME/$DISTRO_NAME.jar.sha1";\
+    apk del .build-deps; \
+    chown -R "$ZOO_USER:$ZOO_USER" "/$DISTRO_NAME"
```
Comment (31z4) [inline 3.5.3/Dockerfile]: Are you sure this is necessary?

File: 3.5.3/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+
+# Download Apache Zookeeper, verify its PGP signature, untar and clean up
+RUN set -ex; \
+    apk add --no-cache --virtual .build-deps \
+        gnupg; \
+    wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz"; \
+    wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz.asc"; \
+    wget -q "http://www.apache.org/dist/zookeeper/KEYS"; \
+    export GNUPGHOME="$(mktemp -d)"; \
+    gpg --import KEYS ; \
```
Comment (31z4) [inline 3.5.3/Dockerfile]: Please remove the space before `;`.

File: 3.5.3/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    gpg --batch --verify "$DISTRO_NAME.tar.gz.asc" "$DISTRO_NAME.tar.gz"; \
+    # TODO: the distribution given by apache is not gzipped correctly - unzip correctly when fixed
+    tar -xf "$DISTRO_NAME.tar.gz"; \
+    mv "$DISTRO_NAME/conf/"* "$ZOO_CONF_DIR"; \
+    rm -rf "$GNUPGHOME" "$DISTRO_NAME.tar.gz" "$DISTRO_NAME.tar.gz.asc"; \
+    rm -rf "/$DISTRO_NAME/CHANGES.txt" \
```
Comment (31z4) [inline 3.5.3/Dockerfile]: It this OK to remove the files from the official distribution? Are you sure this doesn't violate the license?

**Comment by 31z4** [comment]
Hey @duncaan, thanks for your submission! There are few things I noticed during local testing.

1. There is a warning during the build
```
[WARNING]: Empty continuation line found in:
    RUN set -ex;     apk add --no-cache --virtual .build-deps         gnupg;     wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz";     wget -q "http://www.apache.org/dist/zookeeper/$DISTRO_NAME/$DISTRO_NAME.tar.gz.asc";     wget -q "http://www.apache.org/dist/zookeeper/KEYS";     export GNUPGHOME="$(mktemp -d)";     gpg --import KEYS ;     gpg --batch --verify "$DISTRO_NAME.tar.gz.asc" "$DISTRO_NAME.tar.gz";     tar -xf "$DISTRO_NAME.tar.gz";     mv "$DISTRO_NAME/conf/"* "$ZOO_CONF_DIR";     rm -rf "$GNUPGHOME" "$DISTRO_NAME.tar.gz" "$DISTRO_NAME.tar.gz.asc";     rm -rf "/$DISTRO_NAME/CHANGES.txt"       "$DISTRO_NAME/README.txt"       "$DISTRO_NAME/NOTICE.txt"       "$DISTRO_NAME/CHANGES.txt"       "$DISTRO_NAME/README_packaging.txt"       "$DISTRO_NAME/build.xml"       "$DISTRO_NAME/config"       "$DISTRO_NAME/contrib"       "$DISTRO_NAME/dist-maven"       "$DISTRO_NAME/docs"       "$DISTRO_NAME/ivy.xml"       "$DISTRO_NAME/ivysettings.xml"       "$DISTRO_NAME/recipes"       "$DISTRO_NAME/src"       "$DISTRO_NAME/$DISTRO_NAME.jar.asc"       "$DISTRO_NAME/$DISTRO_NAME.jar.md5"       "$DISTRO_NAME/$DISTRO_NAME.jar.sha1";    apk del .build-deps;     chown -R "$ZOO_USER:$ZOO_USER" "/$DISTRO_NAME"
[WARNING]: Empty continuation lines will become errors in a future release.
```
2. When I start the container, there is an exception that's beign logged periodically
```console
$ docker run -it --rm -p 8080:8080 zk:3.5 
ZooKeeper JMX enabled by default
Using config: /conf/zoo.cfg
2017-10-21 05:39:27,407 [myid:] - INFO  [main:QuorumPeerConfig@117] - Reading configuration from: /conf/zoo.cfg
2017-10-21 05:39:27,414 [myid:] - INFO  [main:QuorumPeerConfig@317] - clientPort is not set
2017-10-21 05:39:27,414 [myid:] - INFO  [main:QuorumPeerConfig@331] - secureClientPort is not set
2017-10-21 05:39:27,419 [myid:1] - INFO  [main:DatadirCleanupManager@78] - autopurge.snapRetainCount set to 3
2017-10-21 05:39:27,420 [myid:1] - INFO  [main:DatadirCleanupManager@79] - autopurge.purgeInterval set to 0
2017-10-21 05:39:27,420 [myid:1] - INFO  [main:DatadirCleanupManager@101] - Purge task is not scheduled.
2017-10-21 05:39:27,420 [myid:1] - WARN  [main:QuorumPeerMain@122] - Either no config or no quorum defined in config, running  in standalone mode
2017-10-21 05:39:27,422 [myid:1] - INFO  [main:ManagedUtil@46] - Log4j found with jmx enabled.
2017-10-21 05:39:27,436 [myid:1] - INFO  [main:QuorumPeerConfig@117] - Reading configuration from: /conf/zoo.cfg
2017-10-21 05:39:27,437 [myid:1] - INFO  [main:QuorumPeerConfig@317] - clientPort is not set
2017-10-21 05:39:27,437 [myid:1] - INFO  [main:QuorumPeerConfig@331] - secureClientPort is not set
2017-10-21 05:39:27,438 [myid:1] - INFO  [main:ZooKeeperServerMain@115] - Starting server
2017-10-21 05:39:27,455 [myid:1] - INFO  [main:Environment@109] - Server environment:zookeeper.version=3.5.3-beta-8ce24f9e675cbefffb8f21a47e06b42864475a60, built on 04/03/2017 16:19 GMT
2017-10-21 05:39:27,455 [myid:1] - INFO  [main:Environment@109] - Server environment:host.name=5e2b44a8cf25
2017-10-21 05:39:27,455 [myid:1] - INFO  [main:Environment@109] - Server environment:java.version=1.8.0_131
2017-10-21 05:39:27,455 [myid:1] - INFO  [main:Environment@109] - Server environment:java.vendor=Oracle Corporation
2017-10-21 05:39:27,457 [myid:1] - INFO  [main:Environment@109] - Server environment:java.home=/usr/lib/jvm/java-1.8-openjdk/jre
2017-10-21 05:39:27,458 [myid:1] - INFO  [main:Environment@109] - Server environment:java.class.path=/zookeeper-3.5.3-beta/bin/../build/classes:/zookeeper-3.5.3-beta/bin/../build/lib/*.jar:/zookeeper-3.5.3-beta/bin/../lib/slf4j-log4j12-1.7.5.jar:/zookeeper-3.5.3-beta/bin/../lib/slf4j-api-1.7.5.jar:/zookeeper-3.5.3-beta/bin/../lib/netty-3.10.5.Final.jar:/zookeeper-3.5.3-beta/bin/../lib/log4j-1.2.17.jar:/zookeeper-3.5.3-beta/bin/../lib/jline-2.11.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-util-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-servlet-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-server-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-security-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-io-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/jetty-http-9.2.18.v20160721.jar:/zookeeper-3.5.3-beta/bin/../lib/javax.servlet-api-3.1.0.jar:/zookeeper-3.5.3-beta/bin/../lib/jackson-mapper-asl-1.9.11.jar:/zookeeper-3.5.3-beta/bin/../lib/jackson-core-asl-1.9.11.jar:/zookeeper-3.5.3-beta/bin/../lib/commons-cli-1.2.jar:/zookeeper-3.5.3-beta/bin/../zookeeper-3.5.3-beta.jar:/zookeeper-3.5.3-beta/bin/../src/java/lib/*.jar:/conf:
2017-10-21 05:39:27,458 [myid:1] - INFO  [main:Environment@109] - Server environment:java.library.path=/usr/lib/jvm/java-1.8-openjdk/jre/lib/amd64/server:/usr/lib/jvm/java-1.8-openjdk/jre/lib/amd64:/usr/lib/jvm/java-1.8-openjdk/jre/../lib/amd64:/usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib
2017-10-21 05:39:27,458 [myid:1] - INFO  [main:Environment@109] - Server environment:java.io.tmpdir=/tmp
2017-10-21 05:39:27,459 [myid:1] - INFO  [main:Environment@109] - Server environment:java.compiler=<NA>
2017-10-21 05:39:27,459 [myid:1] - INFO  [main:Environment@109] - Server environment:os.name=Linux
2017-10-21 05:39:27,459 [myid:1] - INFO  [main:Environment@109] - Server environment:os.arch=amd64
2017-10-21 05:39:27,459 [myid:1] - INFO  [main:Environment@109] - Server environment:os.version=4.9.49-moby
2017-10-21 05:39:27,459 [myid:1] - INFO  [main:Environment@109] - Server environment:user.name=zookeeper
2017-10-21 05:39:27,460 [myid:1] - INFO  [main:Environment@109] - Server environment:user.home=/home/zookeeper
2017-10-21 05:39:27,460 [myid:1] - INFO  [main:Environment@109] - Server environment:user.dir=/zookeeper-3.5.3-beta
2017-10-21 05:39:27,460 [myid:1] - INFO  [main:Environment@109] - Server environment:os.memory.free=51MB
2017-10-21 05:39:27,460 [myid:1] - INFO  [main:Environment@109] - Server environment:os.memory.max=889MB
2017-10-21 05:39:27,461 [myid:1] - INFO  [main:Environment@109] - Server environment:os.memory.total=59MB
2017-10-21 05:39:27,465 [myid:1] - INFO  [main:ZooKeeperServer@907] - minSessionTimeout set to 4000
2017-10-21 05:39:27,466 [myid:1] - INFO  [main:ZooKeeperServer@916] - maxSessionTimeout set to 40000
2017-10-21 05:39:27,467 [myid:1] - INFO  [main:ZooKeeperServer@159] - Created server with tickTime 2000 minSessionTimeout 4000 maxSessionTimeout 40000 datadir /datalog/version-2 snapdir /data/version-2
2017-10-21 05:39:27,502 [myid:1] - INFO  [main:Log@186] - Logging initialized @641ms
2017-10-21 05:39:27,576 [myid:1] - WARN  [main:ContextHandler@1339] - o.e.j.s.ServletContextHandler@d2cc05a{/,null,null} contextPath ends with /*
2017-10-21 05:39:27,576 [myid:1] - WARN  [main:ContextHandler@1350] - Empty contextPath
2017-10-21 05:39:27,595 [myid:1] - INFO  [main:Server@327] - jetty-9.2.18.v20160721
2017-10-21 05:39:27,632 [myid:1] - INFO  [main:ContextHandler@744] - Started o.e.j.s.ServletContextHandler@d2cc05a{/,null,AVAILABLE}
2017-10-21 05:39:27,652 [myid:1] - INFO  [main:AbstractConnector@266] - Started ServerConnector@20322d26{HTTP/1.1}{0.0.0.0:8080}
2017-10-21 05:39:27,652 [myid:1] - INFO  [main:Server@379] - Started @795ms
2017-10-21 05:39:27,653 [myid:1] - INFO  [main:JettyAdminServer@112] - Started AdminServer on address 0.0.0.0, port 8080 and command URL /commands
2017-10-21 05:39:27,656 [myid:1] - INFO  [main:ContainerManager@64] - Using checkIntervalMs=60000 maxPerMinute=10000
2017-10-21 05:40:27,659 [myid:1] - ERROR [ContainerManagerTask:ContainerManager$1@84] - Error checking containers
java.lang.NullPointerException
	at org.apache.zookeeper.server.ContainerManager.getCandidates(ContainerManager.java:143)
	at org.apache.zookeeper.server.ContainerManager.checkContainers(ContainerManager.java:112)
	at org.apache.zookeeper.server.ContainerManager$1.run(ContainerManager.java:78)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
2017-10-21 05:41:27,656 [myid:1] - ERROR [ContainerManagerTask:ContainerManager$1@84] - Error checking containers
java.lang.NullPointerException
	at org.apache.zookeeper.server.ContainerManager.getCandidates(ContainerManager.java:143)
	at org.apache.zookeeper.server.ContainerManager.checkContainers(ContainerManager.java:112)
	at org.apache.zookeeper.server.ContainerManager$1.run(ContainerManager.java:78)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
2017-10-21 05:42:27,657 [myid:1] - ERROR [ContainerManagerTask:ContainerManager$1@84] - Error checking containers
java.lang.NullPointerException
	at org.apache.zookeeper.server.ContainerManager.getCandidates(ContainerManager.java:143)
	at org.apache.zookeeper.server.ContainerManager.checkContainers(ContainerManager.java:112)
	at org.apache.zookeeper.server.ContainerManager$1.run(ContainerManager.java:78)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
```

File: 3.5.3/Dockerfile (hunk 4 above)
Comment (duncaan) [reply 3.5.3/Dockerfile]: it is a .tar archive but named as a tar.gz.  
https://issues.apache.org/jira/browse/ZOOKEEPER-2885

File: 3.5.3/Dockerfile (hunk 5 above)
Comment (duncaan) [reply 3.5.3/Dockerfile]: I ran into an issue with the entire directory being owned by a different user.  Might just be my environment though.

**Comment by duncaan** [comment]
@31z4 - the warning in the build is mentioned here: https://github.com/nginxinc/docker-nginx/issues/203 as being a bug in docker.  As for the other error, I'll look soon to see if I can narrow it down.

**Comment by duncaan** [comment]
Also, the exception that you mentioned seems to only happen when there are no zk servers in the config.  If I specify a server, the exception never happens.
`sudo docker run -it --rm -e ZOO_SERVERS="server.1=zoo1.local:2888:3888:participant;2181" zk:3.5`

File: README.md [hunk 8]
```diff
@@ -2,6 +2,7 @@
 
 * `3.3.6`, `3.3` [(3.3.6/Dockerfile)](https://github.com/31z4/zookeeper-docker/blob/master/3.3.6/Dockerfile)
 * `3.4.10`, `3.4`, `latest` [(3.4.10/Dockerfile)](https://github.com/31z4/zookeeper-docker/blob/master/3.4.10/Dockerfile)
+* `3.5.3-beta`, `3.5` [(3.5.3/Dockerfile)](https://github.com/31z4/zookeeper-docker/blob/master/3.5.3/Dockerfile)
```
Comment (31z4) [inline README.md]: The folder name should be equal to the tag.

File: 3.5.3/Dockerfile (hunk 5 above)
Comment (31z4) [reply 3.5.3/Dockerfile]: Was it `mkdir: can't create directory '/zookeeper-3.5.3-beta/bin/../logs': Permission denied`? I think it's worth investigating it a bit to understand what are the logs there and is it useful to make it a volume.

**Comment by 31z4** [comment]
@duncaan according to http://zookeeper.apache.org/doc/trunk/zookeeperReconfig.html#ch_reconfig_upgrade

> Starting with 3.5.0 the clientPort and clientPortAddress configuration parameters should no longer be used. Instead, this information is now part of the server keyword specification ...

So that's the real reason for the `Error checking containers
java.lang.NullPointerException` and why your workaround with setting `ZOO_SERVERS` works. Let's take this into account and set default server keyword value in `docker-entrypoint.sh`.

Also it's probably makes sense to set `standaloneEnabled=false` by default. Please see details by follwing the link above.

**Comment by duncaan** [comment]
I stand by setting the permissions on the extracted directory because inside the container, the user is not known to the system (the user and group show up as those that created the archive).  If we want to also modify the log4j.properties file to adjust where the default log directory is, that might require more work.

https://github.com/apache/zookeeper/blob/branch-3.5/conf/log4j.properties#L24 is the default location.

**Comment by 31z4** [comment]
@duncaan the default log directory is set here https://github.com/apache/zookeeper/blob/branch-3.5/bin/zkEnv.sh#L68-L71 and can be just overriden by setting `ZOO_LOG_DIR` env var. But anyway, I agree with chowning the installation directory. Introducing `ZOO_LOG_DIR` should be a separate change.

**Comment by 31z4** [comment]
Thanks for the awesome contribution @duncaan!

**Pull request merged**